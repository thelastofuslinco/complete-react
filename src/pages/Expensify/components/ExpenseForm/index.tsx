import { ConnectedProps, connect } from 'react-redux'
import { addExpense, editExpense } from '../../../../store'
import Input from '../../../../components/Input'
import { ExpenseModel } from '../../../../models/ExpenseModel'

interface Props extends PropsFromRedux, Partial<ExpenseModel> {
  onClick?: () => void
}

const ExpenseForm = ({
  dispatch,
  description,
  amount,
  note,
  id,
  createdAt,
  onClick
}: Props) => {
  const handleSubmit = (event) => {
    event.preventDefault()
    const { description, amount, note } = event.target

    if (id) {
      dispatch(
        editExpense({
          id,
          description: description.value,
          amount: amount.value,
          note: note.value,
          createdAt
        })
      )
      onClick()
    } else {
      dispatch(
        addExpense({
          description: description.value,
          amount: amount.value,
          note: note.value
        })
      )
    }

    description.value = ''
    amount.value = ''
    note.value = ''
  }

  return (
    <form onSubmit={handleSubmit} className="form">
      <Input name="description" type="text" value={description} />
      <Input name="amount" type="number" value={amount} />
      <Input name="note" value={note} />
      <button type="submit">send</button>
    </form>
  )
}

const conector = connect()

type PropsFromRedux = ConnectedProps<typeof conector>

export default conector(ExpenseForm)
