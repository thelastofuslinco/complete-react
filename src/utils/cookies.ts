export function setCookie(name, value, days) {
  const d = new Date()
  d.setTime(d.getTime() + days * 24 * 60 * 60 * 1000)
  const expires = 'expires=' + d.toUTCString()
  document.cookie = `${name}=${value};${expires};path=/;Secure;HttpOnly;SameSite=Strict`
}

export function getCookie(cookieName) {
  const name = cookieName + '='
  const ca = document.cookie.split(';')

  for (let i = 0; i < ca.length; i++) {
    let c = ca[i]
    while (c.charAt(0) == ' ') {
      c = c.substring(1)
    }
    if (c.indexOf(name) == 0) {
      return c.substring(name.length, c.length)
    }
  }
  return ''
}
