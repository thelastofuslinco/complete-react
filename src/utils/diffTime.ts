export const diffTime = (time) => {
  const date1 = new Date(time).getTime()
  const date2 = new Date().getTime()
  const diff = Math.abs(date2 - date1)
  const diffTime = {
    s: Math.floor(diff / 1000),
    m: Math.floor(diff / (1000 * 60)),
    h: Math.floor(diff / (1000 * 60 * 60))
  }

  if (diffTime.h > 1) {
    return `${diffTime.h} hour ago`
  } else if (diffTime.m > 1) {
    return `${diffTime.m} minutes ago`
  } else {
    return `${diffTime.s} seconds ago`
  }
}
