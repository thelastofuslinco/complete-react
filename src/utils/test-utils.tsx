import { render } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
import expensesSlice from '../store/slices/expensesSlice'

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { expenses: expensesSlice },
      preloadedState
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }

  return {
    store,
    ...render(ui, { wrapper: Wrapper, ...renderOptions })
  }
}
