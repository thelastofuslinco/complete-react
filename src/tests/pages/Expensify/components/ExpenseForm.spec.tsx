import React from 'react'
import { fireEvent, screen } from '@testing-library/react'
import { renderWithProviders } from '../../../../utils/test-utils'
import ExpenseForm from '../../../../pages/Expensify/components/ExpenseForm'
import { expenses } from '../../../fixtures/expenses'

describe('Expense form component', () => {
  test('should render expense form', () => {
    renderWithProviders(<ExpenseForm name="" />)
    const form = screen.getByRole('form', { name: '' })
    expect(form).toMatchSnapshot()
  })

  test('should render expense form with data', () => {
    renderWithProviders(<ExpenseForm name="" {...expenses[0]} />)
    const form = screen.getByRole('form', { name: '' })
    expect(form).toMatchSnapshot()
  })

  test('should change expense form', () => {
    renderWithProviders(<ExpenseForm name="" />)
    const form: HTMLFormElement = screen.getByRole('form', { name: '' })
    fireEvent.change(form, {
      target: {
        amount: { value: 90 },
        description: { value: 'any_description' },
        note: { value: 'any_note' }
      }
    })

    expect(form['amount'].value).toBe(90)
    expect(form['description'].value).toBe('any_description')
    expect(form['note'].value).toBe('any_note')
  })

  test('should change and submit expense', () => {
    renderWithProviders(<ExpenseForm name="" />)
    const form: HTMLFormElement = screen.getByRole('form', { name: '' })
    fireEvent.change(form, {
      target: {
        amount: { value: 90 },
        description: { value: 'any_description' },
        note: { value: 'any_note' }
      }
    })
    fireEvent.submit(form)

    expect(form['amount'].value).toBe('')
    expect(form['description'].value).toBe('')
    expect(form['note'].value).toBe('')
  })

  test('should edit expense', () => {
    const handleClick = jest.fn()
    renderWithProviders(
      <ExpenseForm name="" {...expenses[0]} onClick={handleClick} />
    )
    const form: HTMLFormElement = screen.getByRole('form', { name: '' })
    fireEvent.change(form, {
      target: {
        amount: { value: 90 },
        description: { value: 'any_description' },
        note: { value: 'any_note' }
      }
    })
    fireEvent.submit(form)

    expect(handleClick).toHaveBeenCalledTimes(1)
  })
})
