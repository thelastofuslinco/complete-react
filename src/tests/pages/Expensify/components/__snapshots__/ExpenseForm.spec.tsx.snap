// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense form component should render expense form 1`] = `
<form
  class="form"
  name=""
>
  <input
    class="input undefined"
    name="description"
    type="text"
    value=""
  />
  <input
    class="input undefined"
    name="amount"
    type="number"
    value=""
  />
  <input
    class="input undefined"
    name="note"
    value=""
  />
  <button
    type="submit"
  >
    send
  </button>
</form>
`;

exports[`Expense form component should render expense form with data 1`] = `
<form
  class="form"
  name=""
>
  <input
    class="input undefined"
    name="description"
    type="text"
    value="any_description"
  />
  <input
    class="input undefined"
    name="amount"
    type="number"
    value="822"
  />
  <input
    class="input undefined"
    name="note"
    value="any_description"
  />
  <button
    type="submit"
  >
    send
  </button>
</form>
`;
